javalib/src/main/scala/java/util/AbstractMap.scala

```diff
@@ -0,0 +1,209 @@
+package java.util
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+
+object AbstractMap {
+
+  private def equalsEntry[K, V](entry: Map.Entry[K, V], other: Any): Boolean = {
+    other match {
+      case entry2: Map.Entry[_, _] =>
+        (
+          (entry.getKey == null && entry2.getKey == null) ||
+          entry.getKey.equals(entry2.getKey)
+        ) && (
+            (entry.getValue == null && entry2.getValue == null) ||
+            entry.getValue.equals(entry2.getValue)
+          )
+      case _ => false
+    }
+  }
+
+  private def calcHashCode[K, V](entry: Map.Entry[K, V]): Int = {
+    val keyHash =
+      if (entry.getKey == null) 0
+      else entry.getKey.hashCode
+    val valueHash =
+      if (entry.getValue == null) 0
+      else entry.getValue.hashCode
+
+    keyHash ^ valueHash
+  }
+
+  class SimpleEntry[K, V](key: K, value: V) extends Map.Entry[K, V] with Serializable {
+    private var _key: K = key
+    private var _value: V = value
+
+    def this(entry: Map.Entry[K, V]) =
+      this(entry.getKey, entry.getValue)
+
+    override def equals(o: Any): Boolean =
+      equalsEntry(this, o)
+
+    def getKey(): K = _key
+
+    def getValue(): V = _value
+
+    override def hashCode(): Int = calcHashCode(this)
+
+    def setValue(value: V): V = {
+      _value = value
+      _value
+    }
+  }
+
+  class SimpleImmutableEntry[K, V](key: K, value: V) extends Map.Entry[K, V] with Serializable {
+    def this(entry: Map.Entry[K, V]) =
+      this(entry.getKey, entry.getValue)
+
+    override def equals(o: Any): Boolean =
+      equalsEntry(this, o)
+
+    def getKey(): K = key
+
+    def getValue(): V = value
+
+    override def hashCode(): Int = calcHashCode(this)
+
+    def setValue(value: V): V = throw new UnsupportedOperationException()
+  }
+}
+
+abstract class AbstractMap[K, V]() extends java.util.Map[K, V] {
+  self =>
+
+  @tailrec
+  private def findFirst(i: Iterator[Map.Entry[K, V]])(f: (Map.Entry[K, V]) => Boolean,
+    fi: (Iterator[Map.Entry[K, V]]) => Unit = { _ => () }): Map.Entry[K, V] =
+    if (!i.hasNext) null
+    else {
+      val o = i.next
+      if (f(o)) {
+        fi(i)
+        o
+      } else findFirst(i)(f)
+    }
+
+  @tailrec
+  private def onAll[IK, IV](i: Iterator[Map.Entry[IK, IV]])(f: (Map.Entry[IK, IV]) => Unit): Unit =
+    if (!i.hasNext) return
+    else {
+      f(i.next)
+      onAll(i)(f)
+    }
+
+  def size(): Int = entrySet.size
+
+  def isEmpty(): Boolean = size == 0
+
+  def containsValue(value: Any): Boolean =
+    findFirst(entrySet.iterator)({ x => value.equals(x.getValue) }) != null
+
+  def containsKey(key: Any): Boolean =
+    get(key) != null
+
+  def get(key: Any): V = {
+    val res = findFirst(entrySet.iterator)({ x => key.equals(x.getKey) })
+    if (res == null) res.asInstanceOf[V]
+    else res.getValue
```

Comment:
Ah, that is annoying. But at least write `null.asInstanceOf[V]` instead of `res.asInstanceOf[V]`